Python/bytecodes.c

```diff
@@ -2282,42 +2282,43 @@ dummy_func(
             PyDictObject *dict_o = _PyObject_GetManagedDict(owner_o);
             EXIT_IF(dict_o == NULL);
             assert(PyDict_CheckExact((PyObject *)dict_o));
+#ifdef Py_GIL_DISABLED
+            _PyDict_EnsureSharedOnRead(dict_o);
+#endif
```

Comment:
I think we should de-opt here if the condition doesn't hold. The critical section is escaping so previously validated guards might not hold.

Something like:

```
DEOPT_IF(!_Py_IsOwnedByCurrentThread((PyObject *)dict_o) && !IS_DICT_SHARED(dict_o));
```

Reply:
I think we need to do `DEOPT_IF` because apparently we can't mix and match them. I've updated it to do the ownership check and then mark the dictionary as shared when doing the specialization.